File: HexaProtocols/HexaProtocolBase/src/text_component.rs
use serde_json::json; // Usaremos serde_json para facilitar la creación de JSON

#[derive(Debug, Default)]
pub struct TextComponent {
    pub text: Option<String>,
    pub translate: Option<String>,
    pub with: Option<Vec<TextComponent>>,
    pub extra: Option<Vec<TextComponent>>,
    pub color: Option<String>,
    pub bold: Option<bool>,
    pub italic: Option<bool>,
    pub underlined: Option<bool>,
    pub strikethrough: Option<bool>,
    pub obfuscated: Option<bool>,
    pub font: Option<String>,
    pub insertion: Option<String>,
}

impl TextComponent {
    pub fn new() -> Self {
        TextComponent::default()
    }

    pub fn set_text(&mut self, text: &str) -> &mut Self {
        self.text = Some(text.to_string().replace("&", "§"));
        self
    }

    pub fn set_translate(&mut self, key: &str) -> &mut Self {
        self.translate = Some(key.to_string());
        self
    }

    pub fn add_with(&mut self, component: TextComponent) -> &mut Self {
        if self.with.is_none() {
            self.with = Some(Vec::new());
        }
        self.with.as_mut().unwrap().push(component);
        self
    }

    pub fn add_extra(&mut self, component: TextComponent) -> &mut Self {
        if self.extra.is_none() {
            self.extra = Some(Vec::new());
        }
        self.extra.as_mut().unwrap().push(component);
        self
    }

    pub fn set_color(&mut self, color: &str) -> &mut Self {
        self.color = Some(color.to_string());
        self
    }

    pub fn set_bold(&mut self, bold: bool) -> &mut Self {
        self.bold = Some(bold);
        self
    }

    pub fn set_italic(&mut self, italic: bool) -> &mut Self {
        self.italic = Some(italic);
        self
    }

    pub fn set_underlined(&mut self, underlined: bool) -> &mut Self {
        self.underlined = Some(underlined);
        self
    }

    pub fn set_strikethrough(&mut self, strikethrough: bool) -> &mut Self {
        self.strikethrough = Some(strikethrough);
        self
    }

    pub fn set_obfuscated(&mut self, obfuscated: bool) -> &mut Self {
        self.obfuscated = Some(obfuscated);
        self
    }

    pub fn set_font(&mut self, font: &str) -> &mut Self {
        self.font = Some(font.to_string());
        self
    }

    pub fn set_insertion(&mut self, insertion: &str) -> &mut Self {
        self.insertion = Some(insertion.to_string());
        self
    }

    pub fn to_json(&self) -> serde_json::Value {
        let mut component = json!({});

        if let Some(ref text) = self.text {
            component["text"] = json!(text);
        }

        if let Some(ref translate) = self.translate {
            component["translate"] = json!(translate);
        }

        if let Some(ref with) = self.with {
            component["with"] = json!(with.iter().map(|c| c.to_json()).collect::<Vec<_>>());
        }

        if let Some(ref extra) = self.extra {
            component["extra"] = json!(extra.iter().map(|c| c.to_json()).collect::<Vec<_>>());
        }

        if let Some(ref color) = self.color {
            component["color"] = json!(color);
        }

        if let Some(bold) = self.bold {
            component["bold"] = json!(bold);
        }

        if let Some(italic) = self.italic {
            component["italic"] = json!(italic);
        }

        if let Some(underlined) = self.underlined {
            component["underlined"] = json!(underlined);
        }

        if let Some(strikethrough) = self.strikethrough {
            component["strikethrough"] = json!(strikethrough);
        }

        if let Some(obfuscated) = self.obfuscated {
            component["obfuscated"] = json!(obfuscated);
        }

        if let Some(ref font) = self.font {
            component["font"] = json!(font);
        }

        if let Some(ref insertion) = self.insertion {
            component["insertion"] = json!(insertion);
        }

        component
    }
}

File: HexaUtils/style/text_component.go
package style

import (
	"encoding/json"
	"fmt"
)

// TextComponent represents a Minecraft raw JSON text component
type TextComponent struct {
	Text          string          `json:"text,omitempty"`
	Translate     string          `json:"translate,omitempty"`
	Score         *Score          `json:"score,omitempty"`
	Selector      string          `json:"selector,omitempty"`
	Keybind       string          `json:"keybind,omitempty"`
	NBT           string          `json:"nbt,omitempty"`
	Extra         []TextComponent `json:"extra,omitempty"`
	Color         string          `json:"color,omitempty"`
	Bold          bool            `json:"bold,omitempty"`
	Italic        bool            `json:"italic,omitempty"`
	Underlined    bool            `json:"underlined,omitempty"`
	Strikethrough bool            `json:"strikethrough,omitempty"`
	Obfuscated    bool            `json:"obfuscated,omitempty"`
	Insertion     string          `json:"insertion,omitempty"`
	ClickEvent    *ClickEvent     `json:"clickEvent,omitempty"`
	HoverEvent    *HoverEvent     `json:"hoverEvent,omitempty"`
	Font          string          `json:"font,omitempty"`
}

// Score represents the score of a player or entity in the Minecraft scoreboard
type Score struct {
	Name      string `json:"name"`
	Objective string `json:"objective"`
}

// ClickEvent represents the click event with the associated action and value
type ClickEvent struct {
	Action string `json:"action"`
	Value  string `json:"value"`
}

// HoverEvent represents the hover event with the associated action and content
type HoverEvent struct {
	Action   string      `json:"action"`
	Contents interface{} `json:"contents"`
}

// NewTextComponent creates a new text component with optional text content
func NewTextComponent(text string) *TextComponent {
	return &TextComponent{
		Text: text,
	}
}

// SetColor sets the color of the text component
func (tc *TextComponent) SetColor(color string) *TextComponent {
	tc.Color = color
	return tc
}

// SetBold sets the bold formatting of the text component
func (tc *TextComponent) SetBold(bold bool) *TextComponent {
	tc.Bold = bold
	return tc
}

// SetItalic sets the italic formatting of the text component
func (tc *TextComponent) SetItalic(italic bool) *TextComponent {
	tc.Italic = italic
	return tc
}

// SetUnderlined sets the underlined formatting of the text component
func (tc *TextComponent) SetUnderlined(underlined bool) *TextComponent {
	tc.Underlined = underlined
	return tc
}

// SetStrikethrough sets the strikethrough formatting of the text component
func (tc *TextComponent) SetStrikethrough(strikethrough bool) *TextComponent {
	tc.Strikethrough = strikethrough
	return tc
}

// SetObfuscated sets the obfuscated (glitched) text formatting
func (tc *TextComponent) SetObfuscated(obfuscated bool) *TextComponent {
	tc.Obfuscated = obfuscated
	return tc
}

// SetClickEvent sets the click event for the text component
func (tc *TextComponent) SetClickEvent(action, value string) *TextComponent {
	tc.ClickEvent = &ClickEvent{
		Action: action,
		Value:  value,
	}
	return tc
}

// SetHoverEvent sets the hover event for the text component
func (tc *TextComponent) SetHoverEvent(action string, contents interface{}) *TextComponent {
	tc.HoverEvent = &HoverEvent{
		Action:   action,
		Contents: contents,
	}
	return tc
}

// AddExtra adds an additional component to the "extra" array
func (tc *TextComponent) AddExtra(extra *TextComponent) *TextComponent {
	tc.Extra = append(tc.Extra, *extra)
	return tc
}

// SetScore sets the score display for a player
func (tc *TextComponent) SetScore(name, objective string) *TextComponent {
	tc.Score = &Score{
		Name:      name,
		Objective: objective,
	}
	return tc
}

// SetNBT sets the NBT path for the text component
func (tc *TextComponent) SetNBT(nbt string) *TextComponent {
	tc.NBT = nbt
	return tc
}

// ToJSON marshals the TextComponent into a JSON string
func (tc *TextComponent) ToJSON() (string, error) {
	jsonData, err := json.Marshal(tc)
	if err != nil {
		return "", err
	}
	return string(jsonData), nil
}

// ToMap converts the TextComponent to a map[string]interface{} manually
func (tc *TextComponent) ToMap() map[string]interface{} {
	result := make(map[string]interface{})

	if tc.Text != "" {
		result["text"] = tc.Text
	}
	if tc.Translate != "" {
		result["translate"] = tc.Translate
	}
	if tc.Score != nil {
		result["score"] = map[string]interface{}{
			"name":      tc.Score.Name,
			"objective": tc.Score.Objective,
		}
	}
	if tc.Selector != "" {
		result["selector"] = tc.Selector
	}
	if tc.Keybind != "" {
		result["keybind"] = tc.Keybind
	}
	if tc.NBT != "" {
		result["nbt"] = tc.NBT
	}
	if len(tc.Extra) > 0 {
		var extraList []map[string]interface{}
		for _, extra := range tc.Extra {
			extraList = append(extraList, extra.ToMap())
		}
		result["extra"] = extraList
	}
	if tc.Color != "" {
		result["color"] = tc.Color
	}
	if tc.Bold {
		result["bold"] = tc.Bold
	}
	if tc.Italic {
		result["italic"] = tc.Italic
	}
	if tc.Underlined {
		result["underlined"] = tc.Underlined
	}
	if tc.Strikethrough {
		result["strikethrough"] = tc.Strikethrough
	}
	if tc.Obfuscated {
		result["obfuscated"] = tc.Obfuscated
	}
	if tc.Insertion != "" {
		result["insertion"] = tc.Insertion
	}
	if tc.ClickEvent != nil {
		result["clickEvent"] = map[string]interface{}{
			"action": tc.ClickEvent.Action,
			"value":  tc.ClickEvent.Value,
		}
	}

	if tc.HoverEvent != nil {
		hoverEventMap := map[string]interface{}{
			"action": tc.HoverEvent.Action,
		}

		if contentTc, ok := tc.HoverEvent.Contents.(*TextComponent); ok {
			hoverEventMap["contents"] = contentTc.ToMap()
		} else {
			hoverEventMap["contents"] = tc.HoverEvent.Contents
		}

		result["hoverEvent"] = hoverEventMap

	}

	if tc.Font != "" {
		result["font"] = tc.Font
	}

	return result
}

// convertContent converts interface{} content of HoverEvent to map[string]interface{} or keep as is
func convertContent(content interface{}) interface{} {
	if tc, ok := content.(*TextComponent); ok {
		return tc.ToMap()
	}
	return content
}

func main() {
	// Example of creating a complex JSON text
	tc := NewTextComponent("Hello, world!").
		SetColor("red").
		SetBold(true).
		SetItalic(true).
		SetClickEvent("open_url", "https://example.com").
		SetHoverEvent("show_text", NewTextComponent("Click here!")).
		AddExtra(NewTextComponent(" More text...").SetColor("blue")).
		SetScore("Test", "test_objective")

	// Convert to JSON
	jsonText, err := tc.ToJSON()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	fmt.Println(jsonText)

	// convert to map
	mapText := tc.ToMap()
	fmt.Println(mapText)

	tc2 := NewTextComponent("Hover Test").SetHoverEvent("show_text", NewTextComponent("Click here!").SetColor("blue").SetBold(true))
	mapText2 := tc2.ToMap()
	fmt.Println(mapText2)

	tc3 := NewTextComponent("Complex Hover Test").SetHoverEvent("show_text", NewTextComponent("Click here!").SetColor("blue").SetBold(true).AddExtra(NewTextComponent(" and more")).AddExtra(NewTextComponent(" and another")).SetScore("Test2", "test_objective_2"))
	mapText3 := tc3.ToMap()
	fmt.Println(mapText3)
}
